### PR TITLE
Fetch first available date on a month switch event

### LINKED ISSUE #153: Fetch the first valid date for a layer within a month
Currently, when a user clicks on a month, it requests a layer for the first day of the month. If a tropical storm happens on a day besides the first of the month, clicking on a month results in the mismatched date popup alert. We should modify this so that when a user clicks on a month in the timeline navigation or loads a new hazard layer, the selected layer is displayed for the first event within that month.

For example, if a tropical storm occurred on 23-Jan-2021, and a user clicks on the month, or toggles the layer on, we should fetch the storm on 23-January instead of 1-January.
Screen Shot 2021-04-07 at 13 47 44

### PR DESCRIPTION
**Pull request by JorgeMartinezG**
This PR closes #153 

- Available dates are collected from an already defined function **findAvailableDayInMonth**

available_month

**UPDATE**

- This commit also includes the fetch of the closest date on a layer switch event. This is done, since there is a possibility that no available date might be found within the selected month.

Please check instance in http://homeless-giraffe.surge.sh

### CODE REVIEW

**Comment by dubwalla** [comment]
Looking good from UX perspective. Can we update the alert message to read:

No data was found for the layer '10-day rainfall estimate (mm)' on 2019-04-25. The closest date, 2019-04-21 has been loaded instead.

File: src/components/MapView/DateSelector/index.tsx [hunk 1]
```diff
@@ -53,8 +53,8 @@ function DateSelector({ availableDates = [], classes }: DateSelectorProps) {
     setSelectedDate(moment(stateStartDate));
   }, [stateStartDate]);
   function updateMonth(month: number) {
-    const { startDate, endDate } = getMonthStartAndEnd(month, selectedYear);
-    dispatch(updateDateRange({ startDate, endDate }));
+    const dates = findAvailableDayInMonth(month, selectedYear, availableDates);
+    dispatch(updateDateRange({ startDate: dates[0] }));
```
Comment (Ry-DS) [inline src/components/MapView/DateSelector/index.tsx:57]: @dubwalla do we still not use `endDate` for anything? (Noted [here](https://github.com/WFP-VAM/prism-frontend/blob/0ab5eb16973d5b45dbe55cb2f408a2e4d239abe5/src/context/mapStateSlice/index.ts#L9)). Might be worthwhile to just remove to make things simpler.

File: src/components/MapView/DateSelector/utils.ts [hunk 2]
```diff
@@ -15,6 +15,38 @@ export const months = [
   'Dec',
 ];
 
+/**
+ * Return the closest date from a given list of available dates
```
Comment (Ry-DS) [inline src/components/MapView/DateSelector/utils.ts:19]: This is great! Happy there's a JSDoc, and the typing of availableDates clearly shows its expecting a return type from `getTime` :)

File: src/components/MapView/DateSelector/utils.ts [hunk 3]
```diff
@@ -15,6 +15,38 @@ export const months = [
...
+ * @param date
+ * @param availableDates
+ * @return date as momentjs object
+ */
+export function findClosestDate(
+  date: number,
+  availableDates: ReturnType<Date['getTime']>[],
+) {
+  const dateToCheck = moment(date);
+  let closest: any;
```
Comment (Ry-DS) [inline src/components/MapView/DateSelector/utils.ts:29]: Can we avoid the use of any? Maybe if you use `array.reduce` we could ensure strong typing, and also remove the eslint warning `fp/no-mutation`

**Review by Ry-DS — COMMENTED** [review]
A couple quick fixes and should be good to go!

File: src/components/MapView/DateSelector/index.tsx (hunk 1 above)
Comment (dubwalla) [reply src/components/MapView/DateSelector/index.tsx:57]: Let's keep it. This came up again last week and will need to be implemented soon-ish

File: src/components/MapView/DateSelector/utils.ts (hunk 3 above)
Comment (JorgeMartinezG) [reply src/components/MapView/DateSelector/utils.ts]: @Ry-DS thanks for the input. It is a good idea and changes have been made. Could you please re-check again?

**Comment by ericboucher** [comment]
@JorgeMartinezG let's avoid force-pushed if you can, they make it harder to follow the git history and for others to review just your latest changes ;)

**Review by ericboucher — APPROVED** [review]

**Comment by dubwalla** [comment]
Looks good from my side

**Review by dubwalla — APPROVED** [review]
Good to go

**Pull request merged**